In [32]:
import pandas as pd 

In [33]:
data = pd.read_csv("https://raw.githubusercontent.com/Shreyas3108/house-price-prediction/master/kc_house_data.csv")
data

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [34]:
data.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [35]:
df = data.copy()
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

## importing required libraries 

In [36]:
import numpy as np 
import matplotlib.pyplot as plt 


In [37]:
from sklearn.model_selection import train_test_split

In [38]:
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler

In [39]:
df.shape 

(21613, 21)

In [40]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [41]:
df.describe().T

count          mean           std           min  \
id             21613.0  4.580302e+09  2.876566e+09  1.000102e+06   
price          21613.0  5.400881e+05  3.671272e+05  7.500000e+04   
bedrooms       21613.0  3.370842e+00  9.300618e-01  0.000000e+00   
bathrooms      21613.0  2.114757e+00  7.701632e-01  0.000000e+00   
sqft_living    21613.0  2.079900e+03  9.184409e+02  2.900000e+02   
sqft_lot       21613.0  1.510697e+04  4.142051e+04  5.200000e+02   
floors         21613.0  1.494309e+00  5.399889e-01  1.000000e+00   
waterfront     21613.0  7.541757e-03  8.651720e-02  0.000000e+00   
view           21613.0  2.343034e-01  7.663176e-01  0.000000e+00   
condition      21613.0  3.409430e+00  6.507430e-01  1.000000e+00   
grade          21613.0  7.656873e+00  1.175459e+00  1.000000e+00   
sqft_above     21613.0  1.788391e+03  8.280910e+02  2.900000e+02   
sqft_basement  21613.0  2.915090e+02  4.425750e+02  0.000000e+00   
yr_built       21613.0  1.971005e+03  2.937341e+01  1.900000e+03   
yr_renovated   21613.0  8.440226e+01  4.016792e+02  0.000000e+00   
zipcode        21613.0  9.807794e+04  5.350503e+01  9.800100e+04   
lat            21613.0  4.756005e+01  1.385637e-01  4.715590e+01   
long           21613.0 -1.222139e+02  1.408283e-01 -1.225190e+02   
sqft_living15  21613.0  1.986552e+03  6.853913e+02  3.990000e+02   
sqft_lot15     21613.0  1.276846e+04  2.730418e+04  6.510000e+02   

                        25%           50%           75%           max  
id             2.123049e+09  3.904930e+09  7.308900e+09  9.900000e+09  
price          3.219500e+05  4.500000e+05  6.450000e+05  7.700000e+06  
bedrooms       3.000000e+00  3.000000e+00  4.000000e+00  3.300000e+01  
bathrooms      1.750000e+00  2.250000e+00  2.500000e+00  8.000000e+00  
sqft_living    1.427000e+03  1.910000e+03  2.550000e+03  1.354000e+04  
sqft_lot       5.040000e+03  7.618000e+03  1.068800e+04  1.651359e+06  
floors         1.000000e+00  1.500000e+00  2.000000e+00  3.500000e+00  
waterfront     0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
view           0.000000e+00  0.000000e+00  0.000000e+00  4.000000e+00  
condition      3.000000e+00  3.000000e+00  4.000000e+00  5.000000e+00  
grade          7.000000e+00  7.000000e+00  8.000000e+00  1.300000e+01  
sqft_above     1.190000e+03  1.560000e+03  2.210000e+03  9.410000e+03  
sqft_basement  0.000000e+00  0.000000e+00  5.600000e+02  4.820000e+03  
yr_built       1.951000e+03  1.975000e+03  1.997000e+03  2.015000e+03  
yr_renovated   0.000000e+00  0.000000e+00  0.000000e+00  2.015000e+03  
zipcode        9.803300e+04  9.806500e+04  9.811800e+04  9.819900e+04  
lat            4.747100e+01  4.757180e+01  4.767800e+01  4.777760e+01  
long          -1.223280e+02 -1.222300e+02 -1.221250e+02 -1.213150e+02  
sqft_living15  1.490000e+03  1.840000e+03  2.360000e+03  6.210000e+03  
sqft_lot15     5.100000e+03  7.620000e+03  1.008300e+04  8.712000e+05

## independant feature 

In [42]:
X = df.drop(['price'],axis=1)
X

id             date  bedrooms  bathrooms  sqft_living  \
0      7129300520  20141013T000000         3       1.00         1180   
1      6414100192  20141209T000000         3       2.25         2570   
2      5631500400  20150225T000000         2       1.00          770   
3      2487200875  20141209T000000         4       3.00         1960   
4      1954400510  20150218T000000         3       2.00         1680   
...           ...              ...       ...        ...          ...   
21608   263000018  20140521T000000         3       2.50         1530   
21609  6600060120  20150223T000000         4       2.50         2310   
21610  1523300141  20140623T000000         2       0.75         1020   
21611   291310100  20150116T000000         3       2.50         1600   
21612  1523300157  20141015T000000         2       0.75         1020   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0          5650     1.0           0     0          3      7        1180   
1          7242     2.0           0     0          3      7        2170   
2         10000     1.0           0     0          3      6         770   
3          5000     1.0           0     0          5      7        1050   
4          8080     1.0           0     0          3      8        1680   
...         ...     ...         ...   ...        ...    ...         ...   
21608      1131     3.0           0     0          3      8        1530   
21609      5813     2.0           0     0          3      8        2310   
21610      1350     2.0           0     0          3      7        1020   
21611      2388     2.0           0     0          3      8        1600   
21612      1076     2.0           0     0          3      7        1020   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21608              0      2009             0    98103  47.6993 -122.346   
21609              0      2014             0    98146  47.5107 -122.362   
21610              0      2009             0    98144  47.5944 -122.299   
21611              0      2004             0    98027  47.5345 -122.069   
21612              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21608           1530        1509  
21609           1830        7200  
21610           1020        2007  
21611           1410        1287  
21612           1020        1357  

[21613 rows x 20 columns]

## dependant feature 

In [43]:
y= df['price']
y

0        221900.0
1        538000.0
2        180000.0
3        604000.0
4        510000.0
           ...   
21608    360000.0
21609    400000.0
21610    402101.0
21611    400000.0
21612    325000.0
Name: price, Length: 21613, dtype: float64

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

## dropped the irrelevant feature for prediction

In [48]:
X= X.drop(['id','date'],axis=1)

In [49]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   bedrooms       21613 non-null  int64  
 1   bathrooms      21613 non-null  float64
 2   sqft_living    21613 non-null  int64  
 3   sqft_lot       21613 non-null  int64  
 4   floors         21613 non-null  float64
 5   waterfront     21613 non-null  int64  
 6   view           21613 non-null  int64  
 7   condition      21613 non-null  int64  
 8   grade          21613 non-null  int64  
 9   sqft_above     21613 non-null  int64  
 10  sqft_basement  21613 non-null  int64  
 11  yr_built       21613 non-null  int64  
 12  yr_renovated   21613 non-null  int64  
 13  zipcode        21613 non-null  int64  
 14  lat            21613 non-null  float64
 15  long           21613 non-null  float64
 16  sqft_living15  21613 non-null  int64  
 17  sqft_lot15     21613 non-null  int64  
dtypes: flo

## train - test splitting

In [50]:
X_train , X_test , y_train , y_test = train_test_split(X,y,test_size=0.25)

In [51]:
scaler = StandardScaler()

## fit transform the train data 

In [53]:
scaled_train = scaler.fit_transform(X_train)
scaled_train

array([[-0.40180558, -1.45739079, -0.95441096, ..., -1.04134244,
        -0.94255632, -0.11948865],
       [-0.40180558,  1.79729671,  1.07039611, ..., -1.27670372,
         3.09734347, -0.1494589 ],
       [-0.40180558, -0.15551579, -0.80362746, ..., -0.67760228,
        -0.68098007, -0.37954373],
       ...,
       [ 0.66047898,  0.49542171,  1.87816489, ...,  2.38209437,
         2.05103849, -0.23817805],
       [-0.40180558, -0.48098454, -0.04970993, ..., -0.53495908,
        -0.55019195, -0.31628123],
       [-0.40180558, -1.45739079, -0.95441096, ...,  1.46917788,
        -1.1314725 , -0.13627921]])

## transform the test data 

In [55]:
scaled_test = scaler.transform(X_test)
scaled_test

array([[ 0.66047898,  0.82089046,  0.53188359, ...,  0.39222172,
        -0.60832001, -0.12219681],
       [ 1.72276354,  0.49542171,  0.28416783, ...,  0.07127452,
         0.69956122, -0.04080773],
       [-0.40180558, -0.15551579, -0.02386133, ..., -1.71889765,
        -0.50659591,  0.32573199],
       ...,
       [ 0.66047898,  1.47182796,  2.68593367, ..., -0.99854948,
         2.25448668, -0.22662326],
       [-1.46409014, -1.45739079, -0.43743895, ..., -1.24104292,
        -0.55019195, -0.22600941],
       [ 0.66047898,  0.82089046,  0.77959935, ...,  0.0213494 ,
         1.19364968, -0.11331406]])

In [66]:
scaled_test[0]

array([ 0.66047898,  0.82089046,  0.53188359, -0.07577238, -0.92198769,
       -0.08378769, -0.3059234 ,  0.9099308 , -0.56185553, -0.20779127,
        1.50245099, -0.14042479, -0.20973301, -0.37129955, -0.80905345,
        0.39222172, -0.60832001, -0.12219681])

We have to avoid the data leakage , or we want the robust model 

In [56]:
from sklearn.svm import SVR 

## Model building 

In [57]:
model = SVR()
model.fit(scaled_train,y_train)

SVR()

## Prediction on the scaled data 

In [72]:
pred_SVR = model.predict(scaled_test)
pred_SVR

array([450226.18556288, 450953.12113279, 450402.51742912, ...,
       451241.46345963, 449642.16680807, 451677.16808323])

## mannual prediction

In [68]:
model.predict([[ 0.66047898,  0.82089046,  0.53188359, -0.07577238, -0.92198769,
       -0.08378769, -0.3059234 ,  0.9099308 , -0.56185553, -0.20779127,
        1.50245099, -0.14042479, -0.20973301, -0.37129955, -0.80905345,
        0.39222172, -0.60832001, -0.12219681]])

array([450226.18556189])

In [79]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score


## Mean Squared error 

In [75]:
mse = mean_squared_error(y_test, pred_SVR)
print("Mean Squared Error:", mse)

Mean Squared Error: 137641871516.98196


## R2 score 

In [81]:
r2_score_ = r2_score(y_test, pred_SVR)
r2_score_

-0.05919231375963108

## this is the worst  model since r2 score is negative

## GridSearchCV

In [82]:
from sklearn.model_selection import GridSearchCV

In [83]:
param_grid = {'C':[0.1,1,10,100,50],'gamma':[1,0.5,0.1,0.01,0.001]}

In [85]:
grid = GridSearchCV(SVR(),param_grid, cv= 5,verbose=1,n_jobs=-1)

In [86]:
grid.fit(scaled_train,y_train)

Fitting 5 folds for each of 25 candidates, totalling 125 fits


GridSearchCV(cv=5, estimator=SVR(), n_jobs=-1,
             param_grid={'C': [0.1, 1, 10, 100, 50],
                         'gamma': [1, 0.5, 0.1, 0.01, 0.001]},
             verbose=1)

In [87]:
grid.best_params_

{'C': 100, 'gamma': 0.01}

In [88]:
new_model = SVR(C=100,gamma=0.01)
new_model

SVR(C=100, gamma=0.01)

In [89]:
new_model.fit(scaled_train,y_train)

SVR(C=100, gamma=0.01)

In [91]:
pred_SVR1=new_model.predict(scaled_test)
pred_SVR1

array([450316.38498636, 483613.34598359, 446818.49772424, ...,
       569378.12417731, 418367.01388561, 518602.82985615])

In [92]:
r2_score_ = r2_score(y_test, pred_SVR1)
r2_score_

0.10177088283036384

## RandomizedSearchCV

In [93]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import expon, reciprocal 

In [94]:
# Example parameter grid: Adjust according to your model
param_distributions = {
    'C': reciprocal(1, 1000),
    'gamma': expon(scale=1.0),
    'kernel': ['linear', 'rbf']
}


In [95]:


# Random search of parameters
svr_random = RandomizedSearchCV(estimator=SVR(), param_distributions=param_distributions, 
                                n_iter=100, cv=5, verbose=2, random_state=42, n_jobs=-1)



In [96]:
# Fit the model
# Note: Replace X_train and y_train with your training data
svr_random.fit(scaled_train, y_train)



Fitting 5 folds for each of 100 candidates, totalling 500 fits


RandomizedSearchCV(cv=5, estimator=SVR(), n_iter=100, n_jobs=-1,
                   param_distributions={'C': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x000001384592BF90>,
                                        'gamma': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x00000138452B3490>,
                                        'kernel': ['linear', 'rbf']},
                   random_state=42, verbose=2)

In [97]:
# You can view the best parameters from fitting the random search:
best_parameters = svr_random.best_params_
print(f"Best Parameters: {best_parameters}")

Best Parameters: {'C': 890.620438616168, 'gamma': 0.628789100540856, 'kernel': 'linear'}


In [101]:
new_model_R = SVR(C=890.620438616168, gamma=0.628789100540856, kernel='linear')
new_model_R

SVR(C=890.620438616168, gamma=0.628789100540856, kernel='linear')

In [102]:
new_model_R.fit(scaled_train,y_train)

SVR(C=890.620438616168, gamma=0.628789100540856, kernel='linear')

In [103]:
pred_SVR2=new_model_R.predict(scaled_test)
pred_SVR2

array([ 433348.57810689,  605464.92076599,  582542.9035208 , ...,
       1260596.9553543 ,  426715.86794918,  796951.56778959])

In [104]:
r2_score_ = r2_score(y_test, pred_SVR2)
r2_score_

0.661160949130589

## Conclusion

### 1.Randomized Search CV Performance (R² = 0.661160949130589):

* This score indicates a moderate level of predictive accuracy, where approximately 66.12% of the variance in your target variable is explained by the model.
* This suggests that the Randomized Search CV likely found a set of hyperparameters that are more optimal for your model compared to the base model or other configurations.


### 2.Grid Search CV Performance (R² = 0.10177088283036384):

* This is a low R² score, indicating that only about 10.18% of the variance in your target variable is explained by the model tuned with Grid Search CV.

* It suggests that the hyperparameters identified through Grid Search CV are not as effective for your model, leading to a model that performs significantly worse than the one tuned with Randomized Search CV.

# **DECISION TREE**